File: backend/app/api/schemas/atlas.py
from pydantic import BaseModel, Field

class NodeCriticality(BaseModel):
    node: list[int]
    score: float


class EdgeCriticality(BaseModel):
    edge: list[list[int]]
    score: float


class CriticalitySummary(BaseModel):
    node: list[NodeCriticality]
    edge: list[EdgeCriticality]

class GraphNode(BaseModel):
    id: int
    x: int
    y: int


class GraphPoint(BaseModel):
    x: int
    y: int


class GraphEdge(BaseModel):
    source: int
    target: int
    pixels: list[GraphPoint] = Field(
        default_factory=list
    )
    length: int = 0


class GraphData(BaseModel):
    nodes: list[GraphNode]
    edges: list[GraphEdge]

class GraphSummary(BaseModel):
    nodes: int
    edges: int


class SegmentationSummary(BaseModel):
    model_name: str
    inference_time_ms: float
    image_size: list[int]
    mask_shape: list[int]
    metadata: dict


class SimulationSummary(BaseModel):
    scenario: str

    removed_nodes: list
    removed_edges: list

    original_nodes: int
    original_edges: int

    remaining_nodes: int
    remaining_edges: int

    connected_components: int

    largest_component: int

    critical_node: list | None

class VisualizationSummary(BaseModel):
    segmentation_overlay: str | None = None
    road_mask: str | None = None
    skeleton: str | None = None
    graph: str | None = None
    criticality: str | None = None


class AtlasResponse(BaseModel):

    segmentation: SegmentationSummary

    pixel_graph: GraphSummary

    topology_graph: GraphSummary

    topology_graph_data: GraphData
    
    criticality: CriticalitySummary
    
    resilience: dict

    risk: dict

    simulation: SimulationSummary

    recommendation: str | None = None

    visualizations: VisualizationSummary

File: frontend/src/types/atlas.ts
// ============================================================
// ATLAS API TYPES
// ============================================================

// -----------------------------
// Shared coordinate types
// -----------------------------

export type Coordinate = [number, number];

export type EdgeCoordinate = [
  Coordinate,
  Coordinate
];


// -----------------------------
// Segmentation
// -----------------------------

export interface SegmentationMetadata {
  [key: string]: unknown;
}

export interface SegmentationResult {
  model_name: string;
  inference_time_ms: number;
  image_size: number[];
  mask_shape: number[];
  metadata: SegmentationMetadata;
}


// -----------------------------
// Graph
// -----------------------------

export interface GraphNode {
  id: number;
  x: number;
  y: number;
}

export interface GraphPoint {
  x: number;
  y: number;
}

export interface GraphEdge {
  source: number;
  target: number;
  pixels: GraphPoint[];
  length: number;
}

export interface GraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface GraphStats {
  nodes: number;
  edges: number;
}


// -----------------------------
// Criticality
// -----------------------------

export interface NodeCriticality {
  node: Coordinate;
  score: number;
}

export interface EdgeCriticality {
  edge: EdgeCoordinate;
  score: number;
}

export interface CriticalityResult {
  node: NodeCriticality[];
  edge: EdgeCriticality[];
}


// -----------------------------
// Resilience
// -----------------------------

export interface ResilienceResult {
  critical_node: Coordinate | null;
  connected_components: number;
  largest_component: number;
}


// -----------------------------
// Risk
// -----------------------------

export interface RiskResult {
  ari: number;
  level: string;
  recommendation: string;
}


// -----------------------------
// Simulation
// -----------------------------

export interface SimulationResult {
  scenario: string;

  removed_nodes: Coordinate[];

  removed_edges: EdgeCoordinate[];

  original_nodes: number;
  original_edges: number;

  remaining_nodes: number;
  remaining_edges: number;

  connected_components: number;
  largest_component: number;

  critical_node: Coordinate | null;
}


// -----------------------------
// Visualizations
// -----------------------------

export interface VisualizationResult {
  /**
   * Satellite image with segmentation overlay.
   */
  segmentation_overlay: string | null;

  /**
   * Extracted road mask visualization.
   */
  road_mask: string | null;

  /**
   * Skeletonized road network.
   */
  skeleton: string | null;

  /**
   * Topological graph visualization.
   */
  graph: string | null;

  /**
   * Criticality visualization.
   */
  criticality: string | null;
}


// -----------------------------
// Root Response
// -----------------------------

export interface AtlasResult {
  segmentation: SegmentationResult;

  pixel_graph: GraphStats;

  topology_graph: GraphStats;

  topology_graph_data: GraphData;

  criticality: CriticalityResult;

  resilience: ResilienceResult;

  risk: RiskResult;

  simulation: SimulationResult | null;

  recommendation: string | null;

  visualizations: VisualizationResult;
}